acrg/satellite/oco2.py

```diff
@@ -0,0 +1,736 @@
+        raise KeyError("Unable to apply filter based on quality flag. Input dataset does not contain data variable {0}".format(filter_name)) 
+    
+    # Add attribute describing modification made to original data
+    mod_attr = "quality_filter"
+    ds_new = add_history_attr(ds_new,mod_attr)
+    ds_new.attrs[mod_attr] = "Original data has been filtered using {0} variable to include only flag = {1} (indicates good data).".format(filter_name,flag)
+
+        
+    return ds_new
+
+def oco2_pressure_filter(ds):
+    '''
+    The oco2_pressure_filter function removes all points which have at least one pressure_level undefined (-9999.99).
+    Unable to use these points to calculate relevant pressure levels and difference between levels within the atmosphere.
+    
+    Args:
+        ds (xarray.Dataset) : 
+            OCO2 CO2 Level 2 Data Product file opened with xarray as a Dataset.
+            Should contain "pressure_levels" as a data variable.
+            
+    Returns:
+       ds (xarray.Dataset) : 
+           Filtered Dataset with data points with undefined pressure levels removed. 
+    '''
+    
+    filter_name = "pressure_levels"
+    
+    exclude = -9999.99
+    
+    # Filter based on points where any pressure_levels value is -9999.99
+    try:
+        # Note: adds dimension associated with where condition if not already
+        # present. Should be fine for oco2 data (all variables should have
+        # a first order of "time") but may need to changed for other data.
+        ds_new = ds.where(~(ds[filter_name] == exclude).any(axis=1),drop=True) 
+    except KeyError:
+        raise Exception("Unable to apply filter to {0} to exclude values of {1}. Input dataset does not contain data variable {0}".format(filter_name,exclude)) 
+
+    # Add attribute describing modification made to original data
+    mod_attr = "bad_pressure_filter"
+    ds_new = add_history_attr(ds_new,mod_attr)
+    ds_new.attrs[mod_attr] = "Data points with values of {1} for any of the {0} values have been removed.".format(filter_name,exclude)
+
+    return ds_new
+
+def name_pressure_filter(ds,filters,pressure_NAME=None,columns=["latitude","longitude","time","pressure_levels"],
+                         cutoff=5.0,layer_range=[50.,500.],
+                         pressure_base_dir=name_pressure_directory,pressure_domain=None,max_days=31,
+                         day_template=True,pressure_convert=1/100.):
+    '''
+    The name_pressure_filter function removes any data points from a dataset which does not match the criteria set out
+    by the filters parameter (and associated values).
+    Filter options include:
+        - cutoff based on a percentage difference between NAME and first pressure level
+        - check NAME surface pressure is greater than (i.e. below in height) the second pressure level
+        - check rough extent of the surface layer based on NAME pressure and second pressure level is in a certain range
+    
+    Note:
```

Comment:
Maybe can delete the lines within the Note: (L164-167) as I don't know if they're correct anymore or relevant to OCO-2?